rmw_connext_shared_cpp/src/security_logging.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020 Canonical Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcutils/get_env.h>
+#include <rmw/error_handling.h>
+#include <rmw/qos_profiles.h>
+#include <rmw/types.h>
+
+#include "rmw_connext_shared_cpp/security_logging.hpp"
+
+namespace
+{
+// Environment variable names
+// TODO(security-wg): These are intended to be temporary, and need to be refactored into a proper
+// abstraction.
+const char log_file_variable_name[] = "ROS_SECURITY_LOG_FILE";
+const char log_publish_variable_name[] = "ROS_SECURITY_LOG_PUBLISH";
+const char log_verbosity_variable_name[] = "ROS_SECURITY_LOG_VERBOSITY";
+
+// Connext property names. These are being taken from table 10.1 of
+// https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/manuals/connext_dds/dds_security/RTI_SecurityPlugins_GettingStarted.pdf
+const char log_file_property_name[] = "com.rti.serv.secure.logging.log_file";
+const char distribute_enable_property_name[] = "com.rti.serv.secure.logging.distribute.enable";
+const char verbosity_property_name[] = "com.rti.serv.secure.logging.log_level";
+
+const struct
+{
+  const char * const name;
+  const char * const level;
+} supported_verbosities[] =
+{
+  {"EMERGENCY", "0"},
+  {"ALERT", "1"},
+  {"CRITICAL", "2"},
+  {"ERROR", "3"},
+  {"WARNING", "4"},
+  {"NOTICE", "5"},
+  {"INFORMATIONAL", "6"},
```

Comment:
TBH this has been pretty unexpected and I had to go in the code to figure out what value to put to get this working.
I epected one of the following to work:
- what's in the spec:
  - INFO_LEVEL
  - INFORMATIONAL_LEVEL
- What I saw in most DDS examples (Connext, CoreDX): numbers between 0 and 7
- The ROS severity levels https://github.com/ros2/rcutils/blob/a357b67e7d9f00f8fbfe306491438cd1e8ee5219/rcutils/logging.py#L17

And NONE worked...

I believe we need those to be both more visible, with a more informative error message than:
`failed to set security logging verbosity: 7 is not a supported verbosity`
We need to either accept more values, or pick ones that are already agreed on elsewhere, either the numbers or string matching the

Reply:
Both Fast RTPS and Connext follow the enum in section 9.6 of DDS-Security... mostly. Connext uses the numeric value of the enum in 5.3, but deprecated that in v6 to use the strings instead, sans the "_LEVEL". Fast RTPS uses the "_LEVEL." However, given that:

1. We want all RMWs to work the same way, even given that this is temporary, and
2. We want this to be more of a ROS abstraction than a DDS-level thing

We settled on the more clear and less "go read the spec" ish option. You're right though-- would it help to simply print all the supported verbosities in the error message?

Reply:
Also, I'd be happy to use the ROS verbosities instead of that seems more clear. There aren't quite as many levels though so we'll need to map conservatively. Maybe:

| ROS verbosity | DDS verbosity |
|---|---|
| FATAL | EMERGENCY |
| (also ERROR) | ALERT |
| (also ERROR) | CRITICAL |
| ERROR | ERROR |
| WARN | WARNING |
| INFO | NOTICE |
| DEBUG | INFORMATIONAL |

Reply:
I don't feel very strongly either way as this is temporary, as long as we make it more visible.
For the longer term, we'll need to decide what to put in the XML file exposed to users.
As that XML file aim to have only ROS concepts, maybe we should use ROS verbosity levels in the XML.
So for consistency I'd say let's use that here too.
The mapping in the table above looks good to me :+1:

Reply:
I can't really make my mind on whether we should map or not. I see the benefit of sticking to ROS verbosities but can't yet foresee the cons of having less levels. And the specs does not help much here, from section 9.6:
```
    EMERGENCY_LEVEL,    // System is unusable. Should not continue use.
    ALERT_LEVEL,        // Should be corrected immediately
    CRITICAL_LEVEL,     // A failure in primary application.
    ERROR_LEVEL,        // General error conditions
    WARNING_LEVEL,      // May indicate future error if action not taken.
    NOTICE_LEVEL,       // Unusual, but nor erroneous event or condition.
    INFORMATIONAL_LEVEL, // Normal operational. Requires no action.
    DEBUG_LEVEL
```
I guess either way is fine for now so we should prefer the most user friendly solution and that'd be using ROS verbosities.

Reply:
Alright, just pushed a change to use ROS verbosities, as well as a nicer error message when an unsupported once is selected.